sotodlib/site_pipeline/finalize_focal_plane.py

```diff
@@ -0,0 +1,342 @@
+            logger.info("Loading information from file at %s", obs_id)
+            rset = read_dataset(obs_id, "focal_plane")
+            _aman = rset.to_axismanager(axis_key="dets:readout_id")
+            aman = AxisManager(_aman.dets)
+            aman.wrap(name, _aman)
+        else:
+            logger.info("Loading information from observation %s", obs_id)
+            aman = ctx.get_meta(obs_id, dets=config["context"].get("dets", {}))
+        if name not in aman:
+            logger.warning(
+                "\tNo position_match associated with this observation. Skipping."
+            )
+            continue
+
+        # Put SMuRF band channel in the correct place
+        smurf = AxisManager(aman.dets)
+        smurf.wrap("band", aman[name].band, [(0, smurf.dets)])
+        smurf.wrap("channel", aman[name].channel, [(0, smurf.dets)])
+        aman.det_info.wrap("smurf", smurf)
+
+        if detmap is not None:
+            g3u.add_detmap_info(aman, detmap)
+        have_wafer = "wafer" in aman.det_info
+        if not have_wafer:
+            logger.error("\tThis observation has no detmap results, skipping")
+            continue
+
+        det_ids = aman.det_info.det_id
+        x = aman.det_info.wafer.det_x
+        y = aman.det_info.wafer.det_y
+        pol = aman.det_info.wafer.angle
+        if use_matched:
+            det_ids = aman[name].matched_det_id
+            dm_sort = np.argsort(aman.det_info.det_id)
+            mapping = np.argsort(np.argsort(det_ids))
+            x = x[dm_sort][mapping]
+            y = y[dm_sort][mapping]
+            pol = pol[dm_sort][mapping]
+
+        focal_plane = np.column_stack(
+            (aman[name].xi, aman[name].eta, aman[name].polang, x, y, pol)
+        ).astype(float)
+        out_msk = aman[name].outliers
+        focal_plane[out_msk, :3] = np.nan
+
+        for di, fp in zip(det_ids, focal_plane):
+            try:
+                fp_dict[di].append(fp)
+            except KeyError:
+                fp_dict[di] = [fp]
+
+    if not fp_dict:
+        logger.error("No valid observations provided")
+        sys.exit()
+
+    # Compute the average focal plane while ignoring outliers
+    det_id, focal_plane = _avg_focalplane(fp_dict)
+    measured = focal_plane[:3]
+
+    # Get nominal xi, eta, gamma
+    nominal = get_nominal(focal_plane, config["coord_transform"])
+
+    # Compute transformation between the two nominal and measured pointing
+    if np.isnan(measured[2]).all():
+        logger.warning("No polarization data availible, gammas will be nan")
+        nominal = nominal[:2]
+        measured = measured[:2]
+    affine, shift = get_affine(np.vstack(nominal), np.vstack(measured))
+    scale, shear, rot = decompose_affine(affine)
+    rot = decompose_rotation(rot)
+
+    if np.isclose(scale, np.pi / 180.0).any() or np.isclose(scale, 180.0 / np.pi).any():
+        logger.warning(
+            (
+                "Scale factor looks like a deg/rad conversion."
```

Comment:
In messages like this is helpful to print out the value you triggered on.

In this program though, you should probably print out the scale / shear / rot regardless.